### DISCUSSION TITLE
unraid backup scripts

### DISCUSSION THREAD

**Discussion by dkerlee** [immich-app/immich#8416] [Community Guides]
I added these scrips that seem to work fine in Community Applications > User Scripts.

[postgresql-backup-script.txt](https://github.com/immich-app/immich/files/14816406/postgresql-backup-script.txt)

and the actual pics:
`#!/bin/bash`
`rsync -av --progress /mnt/user/memories/immich/library /mnt/disks/TOSHIBA_EXTERNAL_USB/immich`
`rsync -av --progress /mnt/user/memories/immich/upload /mnt/disks/TOSHIBA_EXTERNAL_USB/immich`
`rsync -av --progress /mnt/user/memories/immich/profile /mnt/disks/TOSHIBA_EXTERNAL_USB/immich`

Then I just look at that external hard drive from time to time to make sure everything is there.

**Comment by DrSpaldo** [comment]
hi @dkerlee , Thanks for posting this. I am also looking at a better/easier way to backup my Immich instance inside Unraid. I am not great with the commands in the docs... 

I had a couple of questions about your script; what is BACKUP_DIR & BACKUP_DIR2 ? I assume the first is where the backup is going.. what is the second for?

Also, what is the reference to nextcloud within the script for?

**Reply by LackesLab** [reply]
Here are 2 Scripts: The first backups all postgres databases (including immich db), the second creates a save copy of the appdata dir:
### Postgres DBs Backup
```
#!/bin/bash

# Check if array is started
ls /mnt/disk[1-9]* 1>/dev/null 2>/dev/null
if [ $? -ne 0 ]
then
   echo "ERROR:  Array must be started before using this script"
   exit
fi


BACKUP_DIR="/mnt/user/Backups/PostgresDatabases/"

databases=`docker exec postgres psql -U postgres -l -t | cut -d'|' -f1 | sed -e 's/ //g' -e '/^$/d'`


for db in $databases; do  if [ "$db" != "postgres" ] && [ "$db" != "template0" ] && [ "$db" != "template1" ] && [ "$db" != "template_postgis" ]; then    
    FOLDER=$BACKUP_DIR/$db

    echo "Backing up $db"
    mkdir -p "$FOLDER"

    FILENAME=${db}_$(date +%Y-%m-%d-%H.%M.%S).sql
    FILEPATH=$FOLDER/$FILENAME
    docker exec postgres pg_dump --clean -h localhost -U postgres -d $db > $FILEPATH
    tar czf $FILEPATH.tgz -C $FOLDER $FILENAME
    rm $FILEPATH

    fi
done

find $BACKUP_DIR/ -type f -name '*.tgz' -mtime +7 -exec rm {} \;
```


### Appdata Backup
```
#!/bin/bash

# Get all running containers and store them in $CONTAINERS
mapfile -t CONTAINERS < <( docker inspect --format='{{.Name}}' $(docker ps -q --no-trunc) | cut -c2- )

# Stop all running containers
for container in "${CONTAINERS[@]}"; do
  echo "Stopping ${container}..."
  docker stop ${container} 1> /dev/null
done

# Backup appdata.
echo "Backing up /mnt/user/appdata/"
rsync -a --delete /mnt/user/appdata/ /mnt/user/Backups/appdata_backup/

# Start all containers that were previously running
for container in "${CONTAINERS[@]}"; do
  echo "Starting ${container}..."
  docker start ${container} 1> /dev/null
done
```

**Reply by DrSpaldo** [reply]
Thanks for the reply @LackesLab . 

Are you running this from the Unraid root console/SSH?

Is this backing up the entire appdata folder or just for Immich? Ie. Is there anything I should go in and change within the scripts?

**Comment by dkerlee** [comment]
> what is BACKUP_DIR & BACKUP_DIR2 ? I assume the first is where the backup is going.. what is the second for?
ahhhh, here it is.
BACKUP_DIR is my external USB drive
BACKUP_DIR2 is just a backup directory on the server itself, the disk pool.

the cp $BACKUP_DIR/xxxxxxxxxxxxxxxx command line copies the one back up to a second location. So I have that postgresbackup in two places: DIR, and DIR2. The external USB and the internal disk pool.

The last line: "# Rotate backups.........."
keeps only the 7 most recent files.

Dude! I asked chatgpt to write this stuff for me, and just changed it a bit. Pretty darn cool.





> Also, what is the reference to nextcloud within the script for?
nextcloud... let's see...
so you see how there are two pieces that say "# Perform the database backup" ?
The first one runs a docker command in a docker image called "PostgreSQL_Immich", which is the Immich postgres database. The second one performs the same sort of database backup in my nextcloud database docker, called "postgresql15"

I realize that muddies the waters a bit as this is immich github. But it's what I got, and might be helpful (or confusing! Which is why you asked!).

**Comment by zaggash** [comment]
I am using dbbackup to backup the db in the immich upload folder.
Then I backup the whole upload folder to an external disk then to the cloud with restic.

**Reply by DrSpaldo** [reply]
> I am using dbbackup to backup the db in the immich upload folder. Then I backup the whole upload folder to an external disk then to the cloud with restic.

@zaggash how are you using dbbackup within Unraid?

**Reply by zaggash** [reply]
You can either install it through the community app or with docker-compose.

**Comment by masterwishx** [comment]
Can we use also zfs replication?
